WordPress/gutenberg
# Build Tooling: Optimize build by spawning worker pool

Previously: #8093

This pull request seeks to try to optimize `npm run build`. The previous effort to convert synchronous scripts to asynchronous equivalents from #8093 was adapted here. The changes here also include a refactoring to front-load files aggregation (i.e. discover all files to build at once, rather than by individual packages).

Ultimately, these changes had minimal impact; the larger impact is by adopting the [`worker-farm` package](https://www.npmjs.com/package/worker-farm), which manages creation of subprocesses and concurrency within those subprocesses.

The net result is a **build time decrease of 56%** (~24s to 10.5s, measured by `time node ./bin/packages/build.js`). The benefit will vary by machine, since `worker-farm` defaults to concurrent workers equal to the number of CPUs.

**Testing Instructions:**

There should be no regressions in the build of packages from `npm run build`.

There should be no regressions in the rebuild of packages from `npm run dev` (though there may be some separate bugginess with rebuilds as addressed in #15219).
